README
NAME

    Test::Moose::More - More tools for testing Moose packages

VERSION

    This document describes version 0.049 of Test::Moose::More - released
    July 30, 2017 as part of Test-Moose-More.

SYNOPSIS

        use Test::Moose::More;
    
        is_class_ok 'Some::Class';
        is_role_ok  'Some::Role';
        has_method_ok 'Some::Class', 'foo';
    
        # ... etc

DESCRIPTION

    This package contains a number of additional tests that can be employed
    against Moose classes/roles. It is intended to replace Test::Moose in
    your tests, and re-exports any tests that it has and we do not, yet.

 Export Groups

    By default, this package exports all test functions. You can be more
    selective, however, and there are a number of export groups (aside from
    the default :all) to help you achieve those dreams!

    :all

      All exportable functions.

    :validate

      "validate_attribute", "validate_class", "validate_role",
      "validate_thing"

TEST FUNCTIONS

 meta_ok $thing

    Tests $thing to see if it has a metaclass; $thing may be the class name
    or instance of the class you wish to check. Passes if $thing has a
    metaclass.

 no_meta_ok $thing

    Tests $thing to see if it does not have a metaclass; $thing may be the
    class name or instance of the class you wish to check. Passes if $thing
    does not have a metaclass.

 does_ok $thing, < $role | \@roles >, [ $message ]

    Checks to see if $thing does the given roles. $thing may be the class
    name or instance of the class you wish to check.

    Note that the message will be taken verbatim unless it contains %s
    somewhere; this will be replaced with the name of the role being tested
    for.

 does_not_ok $thing, < $role | \@roles >, [ $message ]

    Checks to see if $thing does not do the given roles. $thing may be the
    class name or instance of the class you wish to check.

    Note that the message will be taken verbatim unless it contains %s
    somewhere; this will be replaced with the name of the role being tested
    for.

 has_attribute_ok $thing, $attribute_name, [ $message ]

    Checks $thing for an attribute named $attribute_name; $thing may be a
    class name, instance, or role name.

 has_method_ok $thing, @methods

    Queries $thing's metaclass to see if $thing has the methods named in
    @methods.

    Note: This does not include inherited methods; see "has_method" in
    Class::MOP::Class.

 has_no_method_ok $thing, @methods

    Queries $thing's metaclass to ensure $thing does not provide the
    methods named in @methods.

    Note: This does not include inherited methods; see "has_method" in
    Class::MOP::Class.

 has_method_from_anywhere_ok $thing, @methods

    Queries $thing's metaclass to see if $thing has the methods named in
    @methods.

    Note: This does include inherited methods; see "find_method_by_name" in
    Class::MOP::Class.

 has_no_method_from_anywhere_ok $thing, @methods

    Queries $thing's metaclass to ensure $thing does not provide the
    methods named in @methods.

    Note: This does include inherited methods; see "find_method_by_name" in
    Class::MOP::Class.

 method_from_pkg_ok $thing, $method, $orig_pkg

    Given a thing (role, class, etc) and a method, test that it originally
    came from $orig_pkg.

 method_not_from_pkg_ok $thing, $method, $orig_pkg

    Given a thing (role, class, etc) and a method, test that it did not
    come from $orig_pkg.

 method_is_accessor_ok $thing, $method

    Given a thing (role, class, etc) and a method, test that the method is
    an accessor -- that is, it descends from Class::MOP::Method::Accessor.

 method_is_not_accessor_ok $thing, $method

    Given a thing (role, class, etc) and a method, test that the method is
    not an accessor -- that is, it does not descend from
    Class::MOP::Method::Accessor.

 definition_context_ok $meta, \%dc

    Validates the definition context of a metaclass instance. This is a
    strict comparison.

 role_wraps_around_method_ok $role, @methods

    Queries $role's metaclass to see if $role wraps the methods named in
    @methods with an around method modifier.

 role_wraps_before_method_ok $role, @methods

    Queries $role's metaclass to see if $role wraps the methods named in
    @methods with an before method modifier.

 role_wraps_after_method_ok $role, @methods

    Queries $role's metaclass to see if $role wraps the methods named in
    @methods with an after method modifier.

 requires_method_ok $thing, @methods

    Queries $thing's metaclass to see if $thing requires the methods named
    in @methods.

    Note that this really only makes sense if $thing is a role.

 does_not_require_method_ok $thing, @methods

    Queries $thing's metaclass to ensure $thing does not require the
    methods named in @methods.

    Note that this really only makes sense if $thing is a role.

 is_immutable_ok $thing

    Passes if $thing is immutable.

 is_not_immutable_ok $thing

    Passes if $thing is not immutable; that is, is mutable.

 is_pristine_ok $thing

    Passes if $thing is pristine. See "is_pristine" in Class::MOP::Class.

 is_not_pristine_ok $thing

    Passes if $thing is not pristine. See "is_pristine" in
    Class::MOP::Class.

 is_role_ok $thing

    Passes if $thing's metaclass is a Moose::Meta::Role.

 is_class_ok $thing

    Passes if $thing's metaclass is a Moose::Meta::Class.

 is_anon_ok $thing

    Passes if $thing is "anonymous".

 is_not_anon_ok $thing

    Passes if $thing is not "anonymous".

 check_sugar_removed_ok $thing

    Ensures that all the standard Moose sugar is no longer directly
    callable on a given package.

 check_sugar_ok $thing

    Checks and makes sure a class/etc can still do all the standard Moose
    sugar.

 does_metaroles_ok $thing => { $mop => [ @traits ], ... };

    Validate the metaclasses associated with a class/role metaclass.

    e.g., if I wanted to validate that the attribute trait for
    MooseX::AttributeShortcuts is actually applied, I could do this:

        { package TestClass; use Moose; use MooseX::AttributeShortcuts; }
        use Test::Moose::More;
        use Test::More;
    
        does_metaroles_ok TestClass => {
           attribute => ['MooseX::AttributeShortcuts::Trait::Attribute'],
        };
        done_testing;

    This function will accept either class or role metaclasses for $thing.

    The MOPs available for classes (Moose::Meta::Class) are:

    class

    attribute

    method

    wrapped_method

    instance

    constructor

    destructor

    The MOPs available for roles (Moose::Meta::Role) are:

    role

    attribute

    method

    required_method

    wrapped_method

    conflicting_method

    application_to_class

    application_to_role

    application_to_instance

    applied_attribute

    Note! Neither this function nor does_not_metaroles_ok() attempts to
    validate that the MOP type passed in is a member of the above lists.
    There's no gain here in implementing such a check, and a negative to be
    had: specifying an invalid MOP type will result in immediate
    explosions, while it's entirely possible other MOP types will be added
    (either to core, via traits, or "let's subclass Moose::Meta::Class/etc
    and implement something new").

 does_not_metaroles_ok $thing => { $mop => [ @traits ], ... };

    As with "does_metaroles_ok", but test that the metaroles are not
    consumed, a la "does_not_ok".

 attribute_options_ok

    Validates that an attribute is set up as expected; like
    validate_attribute(), but only concerns itself with attribute options.

    Note that some of these options will skip if used against attributes
    defined in a role.

      * -subtest => 'subtest name...'

      If set, all tests run (save the first, "does this thing even have
      this attribute?" test) will be wrapped in a subtest, the name of
      which will be whatever -subtest is set to.

      * is => ro|rw

      Tests for reader/writer options set as one would expect.

      * isa => ...

      Validates that the attribute requires its value to be a given type.

      * does => ...

      Validates that the attribute requires its value to do a given role.

      * builder => '...'

      Validates that the attribute expects the method name given to be its
      builder.

      * default => ...

      Validates that the attribute has the given default.

      * init_arg => '...'

      Validates that the attribute has the given initial argument name.

      * lazy => 0|1

      Validates that the attribute is/isn't lazy.

      * required => 0|1

      Validates that setting the attribute's value is/isn't required.

VALIDATION METHODS

 validate_thing

    Runs a bunch of tests against the given $thing, as defined:

        validate_thing $thing => (
    
            attributes => [ ... ],
            methods    => [ ... ],
            isa        => [ ... ],
    
            # ensures sugar is/is-not present
            sugar      => 0,
    
            # ensures $thing does these roles
            does       => [ ... ],
    
            # ensures $thing does not do these roles
            does_not   => [ ... ],
        );

    $thing can be the name of a role or class, an object instance, or a
    metaclass.

      * -subtest => 'subtest name...'

      If set, all tests run will be wrapped in a subtest, the name of which
      will be whatever -subtest is set to.

      * isa => [ ... ]

      A list of superclasses thing should have.

      * anonymous => 0|1

      Check to see if the class is/isn't anonymous.

      * does => [ ... ]

      A list of roles the thing should do.

      * does_not => [ ... ]

      A list of roles the thing should not do.

      * attributes => [ ... ]

      The attributes list specified here is in the form of a list of names,
      each optionally followed by a hashref of options to test the
      attribute for; this hashref takes the same arguments
      "validate_attribute" does. e.g.:

          validate_thing $thing => (
      
              attributes => [
                  'foo',
                  'bar',
                  baz => { is => 'ro', ... },
                  'bip',
              ],
          );

      * methods => [ ... ]

      A list of methods the thing should have; see "has_method_ok".

      * no_methods => [ ... ]

      A list of methods the thing should not have; see "has_no_method_ok".

      * sugar => 0|1

      Ensure that thing can/cannot do the standard Moose sugar.

      * metaclasses => { $mop => { ... }, ... }

      Validates this thing's metaclasses: that is, given a MOP type (e.g.
      class, attribute, method, ...) and a hashref, find the associated
      metaclass of the given type and invoke "validate_thing" on it, using
      the hashref as options for validate_thing().

      e.g.

          validate_thing 'TestClass' => (
              metaclasses  => {
                  attribute => {
                      isa  => [ 'Moose::Meta::Attribute' ],
                      does => [ 'MetaRole::attribute'    ],
                  },
              },
          );

      ...yields:

          # Subtest: Checking the attribute metaclass, Moose::Meta::Class::__ANON__::SERIAL::1
              ok 1 - TestClass's attribute metaclass has a metaclass
              ok 2 - TestClass's attribute metaclass is a Moose class
              ok 3 - TestClass's attribute metaclass isa Moose::Meta::Attribute
              ok 4 - TestClass's attribute metaclass does MetaRole::attribute
              1..4
          ok 1 - Checking the attribute metaclass, Moose::Meta::Class::__ANON__::SERIAL::1

      Note that validate_class() and validate_role() implement this using
      class_metaclasses and role_metaclasses, respectively.

 validate_role

    The same as validate_thing(), but ensures $thing is a role, and allows
    for additional role-specific tests.

        validate_role $thing => (
    
            required_methods => [ ... ],
    
            # ...and all other options from validate_thing()
        );

      * -compose => 0|1

      When true, attempt to compose the role into an anonymous class, then
      use it to run "validate_class". The options we're given are passed to
      validate_class() directly, except that any required_methods entry is
      removed and its contents pushed onto methods. (A stub method for each
      entry in required_methods will also be created in the new class.)

      e.g.:

          ok 1 - TestRole has a metaclass
          ok 2 - TestRole is a Moose role
          ok 3 - TestRole requires method blargh
          ok 4 - TestRole does TestRole
          ok 5 - TestRole does not do TestRole::Two
          ok 6 - TestRole has method method1
          ok 7 - TestRole has an attribute named bar
          # Subtest: role composed into Moose::Meta::Class::__ANON__::SERIAL::1
              ok 1 - TestRole's composed class has a metaclass
              ok 2 - TestRole's composed class is a Moose class
              ok 3 - TestRole's composed class does TestRole
              ok 4 - TestRole's composed class does not do TestRole::Two
              ok 5 - TestRole's composed class has method method1
              ok 6 - TestRole's composed class has method blargh
              ok 7 - TestRole's composed class has an attribute named bar
              1..7
          ok 8 - role composed into Moose::Meta::Class::__ANON__::SERIAL::1
          1..8

      * -subtest => 'subtest name...'

      If set, all tests run will be wrapped in a subtest, the name of which
      will be whatever -subtest is set to.

      * required_methods => [ ... ]

      A list of methods the role requires a consuming class to supply.

      * before => [ ... ]

      A list of methods the role expects to wrap before, on application to
      a class.

      See "before" in Moose for information on before method modifiers.

      * around => [ ... ]

      A list of methods the role expects to wrap around, on application to
      a class.

      See "around" in Moose for information on around method modifiers.

      * after => [ ... ]

      A list of methods the role expects to wrap after, on application to a
      class.

      See "after" in Moose for information on after method modifiers.

      * role_metaroles => { $mop => [ $role, ... ], ... }

      Checks metaclasses to ensure the given metaroles are applied. See
      "does_metaroles_ok".

      * no_role_metaroles => { $mop => [ $role, ... ], ... }

      Checks metaclasses to ensure the given metaroles are applied. See
      "does_not_metaroles_ok".

      * role_metaclasses => { $mop => { ... }, ... }

      Validates this role's metaclasses: that is, given a MOP type (e.g.
      role, attribute, method, ...) and a hashref, find the associated
      metaclass of the given type and invoke "validate_thing" on it, using
      the hashref as options for validate_thing().

      e.g.

          validate_role 'TestRole' => (
              metaclasses  => {
                  attribute => {
                      isa  => [ 'Moose::Meta::Attribute' ],
                      does => [ 'MetaRole::attribute'    ],
                  },
              },
          );

      ...yields:

          # Subtest: Checking the attribute metaclass, Moose::Meta::Class::__ANON__::SERIAL::1
              ok 1 - TestRole's attribute metaclass has a metaclass
              ok 2 - TestRole's attribute metaclass is a Moose class
              ok 3 - TestRole's attribute metaclass isa Moose::Meta::Attribute
              ok 4 - TestRole's attribute metaclass does MetaRole::attribute
              1..4
          ok 1 - Checking the attribute metaclass, Moose::Meta::Class::__ANON__::SERIAL::1

      Note that validate_class() and validate_role() implement this using
      class_metaclasses and role_metaclasses, respectively.

      * class_metaclasses => { $mop => { ... }, ... }

      As with role_metaclasses, above, except that this option is only used
      if -compose is also specified.

 validate_class

    The same as validate_thing(), but ensures $thing is a class, and allows
    for additional class-specific tests.

        validate_class $thing => (
    
            isa  => [ ... ],
    
            attributes => [ ... ],
            methods    => [ ... ],
    
            # ensures sugar is/is-not present
            sugar      => 0,
    
            # ensures $thing does these roles
            does       => [ ... ],
    
            # ensures $thing does not do these roles
            does_not   => [ ... ],
    
            # ...and all other options from validate_thing()
        );

      * -subtest => 'subtest name...'

      If set, all tests run will be wrapped in a subtest, the name of which
      will be whatever -subtest is set to.

      * immutable => 0|1

      Checks the class to see if it is/isn't immutable.

      * class_metaroles => { $mop => [ $role, ... ], ... }

      Checks metaclasses to ensure the given metaroles are applied. See
      "does_metaroles_ok".

      * no_class_metaroles => { $mop => [ $role, ... ], ... }

      Checks metaclasses to ensure the given metaroles are applied. See
      "does_not_metaroles_ok".

      * class_metaclasses => { $mop => { ... }, ... }

      Validates this class' metaclasses: that is, given a MOP type (e.g.
      role, attribute, method, ...) and a hashref, find the associated
      metaclass of the given type and invoke "validate_thing" on it, using
      the hashref as options for validate_thing().

      e.g.

          validate_class 'TestClass' => (
              metaclasses  => {
                  attribute => {
                      isa  => [ 'Moose::Meta::Attribute' ],
                      does => [ 'MetaRole::attribute'    ],
                  },
              },
          );

      ...yields:

          ok 1 - TestClass has a metaclass
          ok 2 - TestClass is a Moose class
          # Subtest: Checking the attribute metaclass, Moose::Meta::Class::__ANON__::SERIAL::1
              ok 1 - TestClass's attribute metaclass has a metaclass
              ok 2 - TestClass's attribute metaclass is a Moose class
              ok 3 - TestClass's attribute metaclass isa Moose::Meta::Attribute
              ok 4 - TestClass's attribute metaclass does MetaRole::attribute
              1..4
          ok 3 - Checking the attribute metaclass, Moose::Meta::Class::__ANON__::SERIAL::1

 validate_attribute

    validate_attribute() allows you to test how an attribute looks once
    built and attached to a class.

    Let's say you have an attribute defined like this:

        has foo => (
            traits  => [ 'TestRole' ],
            is      => 'ro',
            isa     => 'Int',
            builder => '_build_foo',
            lazy    => 1,
        );

    You can use validate_attribute() to ensure that it's built out in the
    way you expect:

        validate_attribute TestClass => foo => (
    
            # tests the attribute metaclass instance to ensure it does the roles
            -does => [ 'TestRole' ],
            # tests the attribute metaclass instance's inheritance
            -isa  => [ 'Moose::Meta::Attribute' ], # for demonstration's sake
    
            traits   => [ 'TestRole' ],
            isa      => 'Int',
            does     => 'Bar',
            handles  => { },
            reader   => 'foo',
            builder  => '_build_foo',
            default  => undef,
            init_arg => 'foo',
            lazy     => 1,
            required => undef,
        );

    Options passed to validate_attribute() prefixed with - test the
    attribute's metaclass instance rather than a setting on the attribute;
    that is, -does ensures that the metaclass does a particular role (e.g.
    MooseX::AttributeShortcuts), while does tests the setting of the
    attribute to require the value do a given role.

    This function takes all the options "attribute_options_ok" takes, as
    well as the following:

      * -subtest => 'subtest name...'

      If set, all tests run will be wrapped in a subtest, the name of which
      will be whatever -subtest is set to.

SEE ALSO

    Please see those modules/websites for more information related to this
    module.

      * Test::Moose

BUGS

    Please report any bugs or feature requests on the bugtracker website
    https://github.com/RsrchBoy/Test-Moose-More/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR

    Chris Weyl <cweyl@alumni.drew.edu>

CONTRIBUTORS

      * Chad Granum <chad.granum@dreamhost.com>

      * Karen Etheridge <ether@cpan.org>

COPYRIGHT AND LICENSE

    This software is Copyright (c) 2017, 2016, 2015, 2014, 2013, 2012 by
    Chris Weyl.

    This is free software, licensed under:

      The GNU Lesser General Public License, Version 2.1, February 1999



Changes
Revision history for Test-Moose-More

{{$NEXT}}
        * switch from Syntax::Keyword::Junction to List::Util to resolve
        * compatibility issues

0.050     2017-09-20 22:41:41 CDT-0500
	* No code changes, releasing 0.049 as non-trial

0.049-TRIAL  2017-07-30 13:27:51 CDT-0500
	* Add definition_context_ok()

0.048     2017-06-17 23:19:13 CDT-0500
	* Add method_is_accessor_ok(), method_is_not_accessor_ok()
	* Fix POD link to Class::MOP::Method::Accessor
	* Add no_meta_ok()
	* Simplify some of the logic in our _validate_thing_guts()'s attribute
	  loop
	* Skip to the next attribute if current one DNE in _thing_guts()
	  attribute loop
	* Use _thing_name() in _role_wraps(), like everybody else
	* In POD, wrap more things with C<> vs '', etc.

0.047     2017-04-24 22:48:33 CDT-0500
	* Add method_from_pkg_ok(), method_not_from_pkg_ok()
	* Tests which support subtests (typically validate_*()) now do something
	  sensible when '-subtest => 1' is given.

0.046     2017-04-14 22:51:08 CDT-0500
	* Add pristine tests
	* Add has{,_no}_method_ok() TBT tests
	* Add:
		does_not_require_method_ok()
		has_{,no_}method_from_anywhere_ok()

0.045     2017-03-24 20:34:45 CDT-0500
	* Fix a POD command error (=func vs =test)
	* No code changes, just promoting the changes from 0.044 to GA

0.044-TRIAL  2017-03-21 20:34:29 CDT-0500
	* Add metarole/class options to validate_...()
	* add has_no_method_ok(), which does exactly what you think it does
	* add no_methods to validate_thing(), incorporating has_no_method_ok()
	  into validate_*()

0.043     2017-02-14 11:56:50 CST-0600
	* Depend on TAP::SimpleOutput 0.009 for our tests; no code changes.

0.042     2017-02-13 09:54:58 CST-0600
	* Use the subtest bits now in TAP::SimpleOutput
	* No code changes, just the above test tweak.  Overdue release as GA :)

0.041-TRIAL  2016-11-12 20:09:02 CST-0600
	* Added a ':validate' export group.
	* Make -compose subtest messages a little cleaner.
	* add does_metaroles_ok()!  (Not sure why I waited this long...)

0.040-TRIAL  2016-11-10 11:00:24 CST-0600
	* Make more things consistent with naming in tests, e.g. "Class's
	  attribute foo" vs "foo", etc.
	* does_{,not_}ok() now handles metaclasses that don't do roles
	* has_attribute_ok now honors -subtest

0.039-TRIAL  2016-10-10 18:38:44 CDT-0500
	* Use prototypes consistently across our test functions.
	* Moose metaclasses now pass is_class_ok().  *le sigh*
	* Role attributes are now validated w.r.t.:
		+ required
		+ lazy
	* Attribute validation checks are now also tested w.r.t.:
		+ required
		+ lazy

0.038     2016-06-22 13:46:27 CDT-0500
	* Eliminate a warning under v5.22+ (github #12)

0.037     2016-01-31 17:02:52 CST-0600
	* No code changes -- releasing 0.036 as GA

0.036-TRIAL  2016-01-20 20:48:30 CST-0600
	* More test functions now honor "name" overrides, resulting in more
	  sensible output when deeply testing using the validate_*() functions.

0.035     2015-08-26 12:40:38 PDT-0700
	* validate_role() will now perform checks on a role to ensure the role
	  intends to wrap methods on application.

0.034-TRIAL  2015-07-29 23:22:49 PDT-0700
	* validate_role(), validate_class(), validate_thing(), and
	  validate_attribute() now accept '-subtest', causing them to wrap all the
	  tests they run under a subtest.

0.033     2015-07-29 18:25:51 PDT-0700
	* validate_role($role, -compose => 1, ...) will now add a sanity test that
	  the composed subclass actually does $role when composing the arguments
	  to pass to validate_class().

0.032     2015-07-11 22:47:44 PDT-0700
	* Better documentation for most of the validate_*() functions.
	* Add a -compose option to validate_role(), to handle a common pattern
	  automagically.
	* When checking attribute traits, we now attempt to resolve them with the
	  appropriate functions from Moose::Util (so that things like native
	  traits are found with a minimum of pain).

0.031     2015-06-30 18:30:27 PDT-0700
	* We now export validate_thing()
	* Add subroutine parameters to the sugar functions.
	* Add sugar checking to validate_thing() ('sugar => 0|1')

0.030     2015-06-28 01:01:58 PDT-0700
	* Add method modifier checks for roles.

0.029     2015-03-29 15:24:40 PDT-0700
	* No code changes -- 0.028_01 doesn't seem to have broken anything, so
	  release as non-dev.

0.028_01  2015-03-14 21:03:50 PDT-0700
	* Sub::Exporter -> Sub::Exporter::Progressive
	* normalize test names to end with '_ok' (e.g. is_{role,class} ->
	  is_{role,class}_ok)

0.028     2015-03-12 19:28:22 PDT-0700
	* Add is_{,not_}immutable_ok tests
	* Add 'immutable' check to validate_class

0.027     2015-03-10 20:16:38 PDT-0700
	* Drop autobox usage (thanks, @ether!)

0.026     2015-01-18 15:11:22 PST-0800
	* **NO CODE CHANGES** Promote 0.025_01 to 0.026.  For realz this time.
	  *le sigh*

0.025_02  2015-01-18 15:09:18 PST-0800
	* **NO CODE CHANGES** Promote 0.025_01 to 0.026

0.025_01  2014-12-23 23:48:18 PST-0800
	* Handle role attributes in a kinder, gentler way
	* Set our test level correctly when validating attributes

0.025     2014-11-21 11:29:28 PST-0800
	* Merge pull request #13 to handle the latest Test::More alphas.
	  Thanks, @exodist!

0.024     2014-05-14 11:58:47 PDT-0700
	* validate_thing() now does not try to run subtests against attributes it
	  does not find

0.023     2014-01-21 22:39:47 PST-0800
	* fix test error with Moose-2.1200 by adding missing coercions

0.022     2013-11-10 23:16:32 PST-0800
	* ...and add an explicit dependency on TAP::SimpleOutput 0.002

0.021     2013-11-10 23:13:27 PST-0800
	* Handle Test::More's new output for subtests in 0.98_05+
	* Dodge isa_ok()'s output change by rolling our own

0.020     2013-08-01 07:33:57 PDT-0700
	* Drop Perl6::Junction in favor of Syntax::Keyword::Junction Same
	  thing -- essentially, AFAICT -- except that we don't get 'smartmatch
	  operator is experimental' warnings in 5.18 :)

0.019     2013-01-21 07:35:07 PST-0800
	* no code changes, netadata fixup only -- stop requiring

t/funcs.pm

# use as:
my ($_ok, $_nok, $_skip) = counters();

sub counters {
    my $level = shift @_ || 0;
    $level *= 4;
    my $i = 0;

    my $indent = !$level ? q{} : (' ' x $level);

    return (
        sub { $indent .     'ok ' . ++$i . " - $_[0]"      },
        sub { $indent . 'not ok ' . ++$i . " - $_[0]"      },
        sub { $indent .     'ok ' . ++$i . " # skip $_[0]" },
        sub { $indent . "1..$i"                            },
    );
}

!!42;


lib/Test/Moose/More/Utils.pm
#
# This file is part of Test-Moose-More
#
# This software is Copyright (c) 2017, 2016, 2015, 2014, 2013, 2012 by Chris Weyl.
#
# This is free software, licensed under:
#
#   The GNU Lesser General Public License, Version 2.1, February 1999
#
package Test::Moose::More::Utils;
our $AUTHORITY = 'cpan:RSRCHBOY';
$Test::Moose::More::Utils::VERSION = '0.049'; # TRIAL
# ABSTRACT: Various utility functions for TMM (and maybe others!)

use strict;
use warnings;

use Sub::Exporter::Progressive -setup => {

    exports => [ qw{
        get_mop_metaclass_for
        known_sugar
    } ],

    groups  => { default  => [':all'] },
};

use Carp 'croak';
use List::Util 1.33 qw( first all );
use Scalar::Util 'blessed';


sub get_mop_metaclass_for {
    my ($mop, $meta) = @_;

    # FIXME make this less... bad
    # short-circuit!  Better a special case here than *everywhere* else
    return blessed $meta
        if $mop eq 'class' || $mop eq 'role';

    # this code largely lifted from Moose::Util::MetaRole
    my $attr =
        first { $_ }
        map { $meta->meta->find_attribute_by_name($_) }
        ("${mop}_metaclass", "${mop}_class")
        ;

    croak "Cannot find attribute storing the metaclass for $mop in " . $meta->name
        unless $attr;

    my $read_method = $attr->get_read_method;

    return $meta->$read_method();
}


sub known_sugar () { qw{ has around augment inner before after blessed confess } }

!!42;

__END__

=pod

=encoding UTF-8

=for :stopwords Chris Weyl Chad Etheridge Granum Karen TMM

=head1 NAME

Test::Moose::More::Utils - Various utility functions for TMM (and maybe others!)

=head1 VERSION

This document describes version 0.049 of Test::Moose::More::Utils - released July 30, 2017 as part of Test-Moose-More.

=head1 FUNCTIONS

=head2 get_mop_metaclass_for $mop => $meta

Given a MOP name (e.g. attribute), rummage through $meta (a metaclass) to reveal the MOP's metaclass.

e.g.

    get_metaclass_for attribute => __PACKAGE__->meta;

=head2 known_sugar

Returns a list of all the known standard Moose sugar (has, extends, etc).

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<Test::Moose::More|Test::Moose::More>

=item *

L<L<Moose::Util::MetaRole> -- for much of the "find the metaclass for X mop" code|L<Moose::Util::MetaRole> -- for much of the "find the metaclass for X mop" code>

=back

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
L<https://github.com/RsrchBoy/Test-Moose-More/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017, 2016, 2015, 2014, 2013, 2012 by Chris Weyl.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 2.1, February 1999

=cut


Makefile.PL
#
# This file is part of Test-Moose-More
#
# This software is Copyright (c) 2017, 2016, 2015, 2014, 2013, 2012 by Chris Weyl.
#
# This is free software, licensed under:
#
#   The GNU Lesser General Public License, Version 2.1, February 1999
#
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.010.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "More tools for testing Moose packages",
  "AUTHOR" => "Chris Weyl <cweyl\@alumni.drew.edu>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-Moose-More",
  "LICENSE" => "lgpl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Test::Moose::More",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::OptList" => 0,
    "List::MoreUtils" => 0,
    "List::Util" => "1.45",
    "Moose::Util" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "Scalar::Util" => 0,
    "Sub::Exporter::Progressive" => 0,
    "Syntax::Keyword::Junction" => 0,
    "Test::Builder" => 0,
    "Test::Moose" => 0,
    "Test::More" => "0.94",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "List::Util" => "1.45",
    "Moose" => 0,
    "Moose::Meta::Attribute" => 0,
    "Moose::Role" => 0,
    "Moose::Util::MetaRole" => 0,
    "TAP::SimpleOutput" => "0.009",
    "Test::Builder::Tester" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.94",
    "blib" => "1.01",
    "namespace::autoclean" => 0
  },
  "VERSION" => "0.049",
  "test" => {
    "TESTS" => "t/*.t t/attribute_options_ok/*.t t/is_class_ok/*.t t/validate_attribute/*.t t/validate_class/*.t t/validate_role/*.t t/validate_thing/*.t t/wrapped/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::OptList" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::MoreUtils" => 0,
  "List::Util" => "1.45",
  "Moose" => 0,
  "Moose::Meta::Attribute" => 0,
  "Moose::Role" => 0,
  "Moose::Util" => 0,
  "Moose::Util::MetaRole" => 0,
  "Moose::Util::TypeConstraints" => 0,
  "Scalar::Util" => 0,
  "Sub::Exporter::Progressive" => 0,
  "Syntax::Keyword::Junction" => 0,
  "TAP::SimpleOutput" => "0.009",
  "Test::Builder" => 0,
  "Test::Builder::Tester" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::Moose" => 0,
  "Test::More" => "0.94",
  "blib" => "1.01",
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
